File: sol/815.py
from collections import defaultdict, deque

class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        if source == target:
            return 0
        routes = list(frozenset(e) for e in routes if len(e) >= 2)
        n = len(routes)
        edgeDict = defaultdict(list)
        h = deque()
        endRoute = set()
        for i in range(n):
            for j in range(i + 1, n):
                if not routes[i].isdisjoint(routes[j]):
                    edgeDict[i].append(j)
                    edgeDict[j].append(i)
            if source in routes[i]:
                h.append((1, i))
            if target in routes[i]:
                endRoute.add(i)
        if len(endRoute) == 0:
            return -1
        if any(pt[1] in endRoute for pt in h):
            return 1
        visited = set()
        while len(h) != 0:
            currCost, currRoute = h.popleft()
            if currRoute in visited:
                continue
            visited.add(currRoute)
            for nei in edgeDict[currRoute]:
                if nei not in visited:
                    if nei in endRoute:
                        return currCost + 1
                    h.append((currCost + 1, nei))
        return -1

File: sol/815/815.cpp
#include <queue>
#include <unordered_set>

struct Solution {
    inline bool hasIntersection(const unordered_set<int> &s1, const unordered_set<int> &s2){
        for (const int e : s1){
            if (s2.contains(e)){
                return true;
            }
        }
        return false;
    }

    int numBusesToDestination(const vector<vector<int>>& routes, int source, int target) {
        if (source == target){
            return 0;
        }
        vector<unordered_set<int>> routeSet(routes.size()), edgeDict(routes.size());
        queue<pair<int, int>> q;
        unordered_set<int> endRoutes, visited;
        for (int i = 0; i < routes.size(); ++i){
            routeSet[i].insert(routes[i].cbegin(), routes[i].cend());
            for (int j = 0; j < i; ++j){
                if (this->hasIntersection(routeSet[i], routeSet[j])){
                    edgeDict[i].insert(j);
                    edgeDict[j].insert(i);
                }
            }
            if (routeSet[i].contains(source)){
                q.push(make_pair(1, i));
            }
            if (routeSet[i].contains(target)){
                endRoutes.insert(i);
            }
        }
        if (endRoutes.empty()){
            return -1;
        }
        while (!q.empty()){
            auto p = q.front();
            q.pop();
            if (visited.contains(p.second)){
                continue;
            }
            if (endRoutes.contains(p.second)){
                return p.first;
            }
            visited.insert(p.second);
            for (const auto nei : edgeDict[p.second]){
                if (!visited.contains(nei)){
                    q.push(make_pair(p.first + 1, nei));
                }
            }
        }
        return -1;
    }
};
